daemon/src/model.rs

```diff
@@ -692,17 +690,145 @@ impl str::FromStr for OpeningFee {
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         let amount_sat: u64 = s.parse()?;
-        Ok(OpeningFee { amount_sat })
+        Ok(OpeningFee {
+            fee: Amount::from_sat(amount_sat),
+        })
     }
 }
 
+impl_sqlx_type_display_from_str!(OpeningFee);
+
 impl Default for OpeningFee {
     fn default() -> Self {
-        Self::new(Amount::ZERO)
+        Self { fee: Amount::ZERO }
     }
 }
 
-impl_sqlx_type_display_from_str!(OpeningFee);
+impl From<OpeningFee> for Amount {
+    fn from(opening_fee: OpeningFee) -> Self {
+        opening_fee.fee
+    }
+}
+
+/// Funding fee as defined by position and rate
+///
+/// Position and rate define if the funding rate is to be added or subtracted when summing up the
+/// fees.
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct FundingFee {
+    #[serde(with = "bdk::bitcoin::util::amount::serde::as_sat")]
+    fee: Amount,
+    rate: FundingRate,
+}
+
+impl FundingFee {
+    pub fn new(fee: Amount, rate: FundingRate) -> Self {
+        Self { fee, rate }
+    }
+
+    pub fn into_signed_amount(self, position: Position) -> SignedAmount {
+        let fee: i64 = self.fee.as_sat().try_into().expect("not to overflow");
+
+        let signed = if (position == Position::Long && self.rate.0.is_sign_positive())
+            || (position == Position::Short && self.rate.0.is_sign_negative())
+        {
+            fee
+        } else {
+            -fee
+        };
+
+        SignedAmount::from_sat(signed)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum FeeFlow {
```

Comment:
> I wonder if we could model this differently: Instead of having LongPaysShort and vise-verca we have
FeeFlow::Incoming and FeeFlow::Outgoing

Interesting and possible, but not sure it actually makes it easier to understand. `Incoming` and `Outgoing` are quite high level again which makes it harder to interpret than `LongPaysShort` and `ShortPaysLong`.

> Maybe we don't have to deal with a SignedAmount and back and forth conversion neither.

How would you achieve that? The accumulated fees are the sum of negative *or* positive fees. I don't see a possibility  to come up with the actual value without being able to subtract and add to the data type. We could theoretically create a datatype that turns an underflow into an amount with a sign, but I don't think that would make it easier to understand.

I think in the end it should always about the position - the fact that we have the opening fee that always goes from taker to maker could be seen as implementation detail. I would hope that this can actually be depicted as a separate output on `lock` at some point, so we don't have to care about it when dealing with the payout curve.

--- 

That said: What do you think @thomaseizinger and @klochowicz? 
I can change this, but like the current version slightly better.

Reply:
The primary driver of this design was to create an _objective_ view onto how the fees need to be paid. `Incoming` and `Outgoing` are relative to the user. The main user of this API is the payout curve which doesn't care about the parties but only about long and short. That is where we need to make the main calculation so we opted to make the interpretation of the value there as easy as possible.

With `Incoming` and `Outgoing`, we wouldn't even be able to reason about the amount in the payout curve without passing in more information like role & position at which point you are just re-writing what is in settle to figure out how much you need to deduct from long and add to short or vice versa.

Reply:
Independent of that, the whole fee market in CFDs seems to talk about fees being paid from long to shorts or vice versa so this seems like a natural modelling of this problem?

Reply:
No worries, it's just an idea because I had a hard time following the logic of `LongPaysShort` vs `ShortPaysLong` and the meaning I personally attach to these terms: _long pays short_ is to be expected in a bull market while _short pays long_ is to be expected in a bear market. Having these enum variants here confuses me as the can be different to each position. 


regarding no `SignedAmount`: 
My thinking was that the value is always positive but it is either _incoming_ (i.e. you get paid) or _outgoing_ (i.e. you have to pay). Based on this you can tell if you have to add or subtract it.

Reply:
> regarding no SignedAmount:
> My thinking was that the value is always positive but it is either incoming (i.e. you get paid) or outgoing (i.e. you have to pay). Based on this you can tell if you have to add or subtract it.

`FundingFee` and `OpeningFee` are depicted as always positive.
But for the `FeeAccount` we have to come up with the *sum* of the fees. We can create a construct where we keep the sum always positive and add something that defines the "sign", but I am not sure it makes it easier. I think in the end the sum calculation would still require a signed amount, but we could store the balance as absolute and add the flow to the internal representation. Is that what you mean?

Reply:
> No worries, it's just an idea because I had a hard time following the logic of `LongPaysShort` vs `ShortPaysLong` and the meaning I personally attach to these terms: _long pays short_ is to be expected in a bull market while _short pays long_ is to be expected in a bear market. Having these enum variants here confuses me as the can be different to each position.

Correct, that is what is depicted here. In a bull market, settlement of the fee account will result in `LongPaysShort`. In a bear market, it will be `ShortPaysLong`.

If a CFD is kept open for long enough, the fees might swing: Going from `LongPaysShort` to `ShortPaysLong`. The opening fee always flows from maker to taker, so depending on the their position, either long or short starts with a credit or debit in fees.

What we've modeled here is the overall accounting of fees in a CFD. In general, that will follow the bull/bear market rule but because we have an opening fee as well (and maybe in the future a closing fee), we also need to take that into account.

> (note: I kept the balance as `SignedAmount` so i can use it directly in the profit calc, but yeah, it's a bit weird that we use it directly there and for the payout curve we use `settle`)

Tracking the balance as absolute is correct from what I understand. The difficult part is the reasoning about what the sign means. Basic arithmetic defines that adding a negative value is subtraction for example. Because of this, handling signed values can be quite hard to reason about. In the payout curve, we only deal with unsigned values because we are splitting a pot of money into two parts. The assumption was that is easier to understand to pass an unsigned amount to that and separately specify, how the fees are meant to be handled (short to long or long to short).

IMO, we should also use `settle` in the P/L calc and simply also use our own position and role there to reason what that means for us.